src/txmempool.cpp

```diff
@@ -1150,3 +1227,169 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
           cachedInnerUsage);
 }
 
+bool CCoinsViewMemPool::IsQualityValid(const CScCertificate& cert, CAmount certFee) const
+{
+    if (!base->IsQualityValid(cert, certFee))
+    {
+        LogPrint("cert", "%s:%s():%d - cert %s NOK\n", __FILE__, __func__, __LINE__, cert.GetHash().ToString());
+        return false;
+    }
+
+    const uint256& scId = cert.GetScId();
+
+    if ((mempool.mapSidechains.count(scId) != 0) &&
+        (!mempool.mapSidechains.at(scId).mBackwardCertificates.size() == 0))
+    {
+
+        for (const auto& entry : mempool.mapSidechains.at(scId).mBackwardCertificates)
+        {
+            const uint256& memPoolCertHash = entry.second;
+            const CScCertificate& memPoolCert = mempool.mapCertificate.at(memPoolCertHash).GetCertificate();
+
+            // should be nothing but the same epoch
+            if (memPoolCert.epochNumber != cert.epochNumber)
+            {
+                LogPrint("cert", "%s:%s():%d - WARNING: cert %s epoch=%d vs cert %s epoch=%d\n", __FILE__,__func__, __LINE__,
+                    cert.GetHash().ToString(), cert.epochNumber, memPoolCertHash.ToString(), memPoolCert.epochNumber);
+                continue;
+            }
+
+            // first of all compare quality
+            if (memPoolCert.quality != cert.quality)
+            {
+                LogPrint("cert", "%s:%s():%d - cert %s q=%d OK; q=%d for same sc/epoch is in mempool\n",
+                    __FILE__,__func__, __LINE__, cert.GetHash().ToString(), cert.quality, memPoolCert.quality);
+            }
+            else
+            {
+                // in case of equal quality, compare fee
+                CAmount memPoolCertFee = mempool.mapCertificate.at(memPoolCertHash).GetFee();
+                if (memPoolCertFee >= certFee)
+                {
+                    LogPrint("cert", "%s:%s():%d - cert %s q=%d, fee=%s NOK; q=%d, fee=%s for same sc/epoch is in mempool\n",
+                        __FILE__, __func__, __LINE__, cert.GetHash().ToString(), cert.quality, FormatMoney(certFee),
+                        memPoolCert.quality, FormatMoney(memPoolCertFee));
+                    return false;
+                }
+            }
+        }
+    }
+    LogPrint("cert", "%s:%s():%d - cert %s q=%d OK\n", __FILE__,  __func__, __LINE__, cert.GetHash().ToString(), cert.quality);
+    return true;
```

Comment:
So if I understand right, this function checks for quality/fee relationship only, but not for vicious spending dependencies?  Maybe it is the right thing to do, but please add comment about it

Reply:
An alternative would be to skip also the check on fees here and to it later.
We would have a cert for which `checkQuality` is true as long as quality is sufficiently high.
Later on, when checking which, of the possible two same quality certificates to keep in mempool, we would look at fees and spending dependencies.
This function could be renamed like `IsQualityHighEnough(cert)` and you would have a much cleaner interface, avoiding the useless CAmount in CCoinsViewCache::IsQualityValid

Reply:
Superseeded by other changes